Source/Dafny/Translator.cs

```diff
@@ -94,6 +96,229 @@ public string FreshNumericId(string prefix = "")
     }
   }
 
+  class SingularDriver
+  {
+    static uint queryCount = 0;
+
+    uint tempVarCount = 0;
+    static string tempVarPrefix = "tpv_";
+
+    List<string> polys;
+    HashSet<string> variables;
+    Dictionary<string, string> opaqueExprs;
+
+    TextWriter writer;
+
+    string AllocateFreshTemp() {
+      var name = String.Format("{0}{1}", SingularDriver.tempVarPrefix, this.tempVarCount);
+      this.tempVarCount += 1;
+      this.variables.Add(name);
+      return name;
+    }
+
+    string OpacifySubExpr(Expression expr) {
+      if (expr is SeqSelectExpr) {
+        var seqSelExpr = expr as SeqSelectExpr;
+        if (!seqSelExpr.SelectOne) {
+          Console.WriteLine("unhandled seqSelExpr");
+          throw new cce.UnreachableException();
+        }
+        var seq = this.OpacifySubExpr(seqSelExpr.Seq);
+        var index = this.OpacifySubExpr(seqSelExpr.E0);
+        return String.Format("{0}[{1}]", seq, index);
```

Comment:
It seems like this is going to emit something like `x[5]` to Singular.  Won't that confuse it?

Reply:
Oh wait, I see.  The goal is just to create a unique string 'key' to use later in the `opaqueExprs` dictionary?

Reply:
Ah yes, exactly. We create a temp variable in place of the expression, and when the exact same expression shows up again, we will just use the temp variable. Singular will have next to no understanding of the expression, only knowing that is some integer.